sql: set volatility of casts in castMap

mgartner opened the pull request:
#### sql: add volatility hint to castMap

This commit adds the `volatilityHint` field to `castMap` and populates
the field based on the hints already in `castInfo`.

Release note: None

#### sql: add intervalStyleAffected and dateStyleAffected to castMap

This commit adds `intervalStyleAffected` and `dateStyleAffected` fields
to `castMap` and sets them based on the behavior of the existing
functions `isIntervalStyleCastAffected` and `isDateStyleCastAffected`.

Release note: None

#### sql: set volatility of casts in castMap

Some of the casts are marked differently than they are in Postgres in
order to match the current volatility in castInfo. For example,
TEXT->INTERVAL is marked as immutable, but it is stable in Postgres:

    postgres=# CREATE TABLE t (a TEXT);
    CREATE TABLE

    postgres=# CREATE INDEX ON t ((a::INTERVAL));
    ERROR:  42P17: functions in index expression must be marked IMMUTABLE

I've left TODOs to mark which cast volatilities do not match Postgres.

Release note: None

#### sql: add casts for unknown and enum types to lookupCast

Release note: None

#### sql: adds casts between string types and array types to castMap

Casts between string types and array types have been added to `castMap`.
They are always stable casts.

Release note: None

#### sql: add casts from numeric types to intervals to castMap

We support casts from INT2, INT4, INT8, FLOAT4, FLOAT8, and DECIMAL to
INTERVAL even though they are not supported by Postgres. This commit
adds these casts to `castMap` so we do not break existing functionality
when `castInfo` is removed.

Release note: None

#### sql: remove castInfo

`castMap` now contains all volatility information and `castInfo` is no
longer needed. It has been removed.

Release note: None

mgartner committed:
sql: add volatility and counter to cast map

This commit adds `volatility` and `counter` fields to `castMap`, which
takes precedence over the entries in `castInfo`. This allows for more
granular configuration of casts. For now, none of the volatilities have
been set in `castMap`, but they will be in future commits. Once all cast
volatilities and volatility hints have been set in `castMap`, we will
remove `castInfo`.

Release note: None

mgartner committed:
sql: set volatilities of OID and REG* types in castMap

Casting between string types and REG* types is a stable cast, while
casting between string types and OID is immutable. This is now correctly
reflected in the `volatility` field of `castMap`.

Fixes #74216

Release note (bug fix): A bug has been fixed that incorrectly allowed
creating computed column expressions, expression indexes, and partial
index predicate expressions with mutable casts between string types and
the types REGCLASS, REGNAMESPACE, REGPROC, REGPROCEDURE, REGROLE, and
REGTYPE. Creating such computed columns, expression indexes, and partial
indexes is now prohibited. Any tables with these types of expressions
may be corrupt and should be dropped and recreated.

mgartner committed:
sql: add volatility hint to castMap

This commit adds the `volatilityHint` field to `castMap` and populates
the field based on the hints already in `castInfo`.

Release note: None

mgartner committed:
sql: add intervalStyleAffected and dateStyleAffected to castMap

This commit adds `intervalStyleAffected` and `dateStyleAffected` fields
to `castMap` and sets them based on the behavior of the existing
functions `isIntervalStyleCastAffected` and `isDateStyleCastAffected`.

Release note: None

mgartner committed:
sql: set volatility of casts in castMap

Some of the casts are marked differently than they are in Postgres in
order to match the current volatility in castInfo. For example,
TEXT->INTERVAL is marked as immutable, but it is stable in Postgres:

    postgres=# CREATE TABLE t (a TEXT);
    CREATE TABLE

    postgres=# CREATE INDEX ON t ((a::INTERVAL));
    ERROR:  42P17: functions in index expression must be marked IMMUTABLE

I've left TODOs to mark which cast volatilities do not match Postgres.

Release note: None

mgartner committed:
sql: add casts for unknown and enum types to lookupCast

Release note: None

mgartner committed:
sql: remove castInfo

`castMap` now contains all volatility information and `castInfo` is no
longer needed. It has been removed.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74296)
<!-- Reviewable:end -->

mgartner committed:
sql: adds casts between string types and array types to castMap

Casts between string types and array types have been added to `castMap`.
They are always stable casts.

Release note: None

mgartner committed:
volatility wip

mgartner committed:
add volatility test
